wms/task_mq_assembly.py

```diff
@@ -0,0 +1,85 @@
+"""The daemon task that 'assembles' message queues via the MQS."""
+
+
+import asyncio
+import logging
+
+from motor.motor_asyncio import AsyncIOMotorClient
+from pymongo import ASCENDING, DESCENDING
+from rest_tools.client import RestClient
+
+from . import database as db
+from .config import ENV
+from .schema.enums import TaskforcePhase
+
+LOGGER = logging.getLogger(__name__)
+
+
+async def startup(mongo_client: AsyncIOMotorClient) -> None:  # type: ignore[valid-type]
+    """Start up the daemon task."""
+    LOGGER.info("Starting up task_mq_assembly...")
+
+    task_directives_client = db.client.WMSMongoClient(
+        mongo_client,
+        db.utils.TASK_DIRECTIVES_COLL_NAME,
+    )
+    taskforces_client = db.client.WMSMongoClient(
+        mongo_client,
+        db.utils.TASKFORCES_COLL_NAME,
+    )
+    mqs_rc = RestClient(
+        ENV.MQS_ADDRESS,
+        logger=logging.getLogger(f"{LOGGER.name}.mqs"),
+    )
+
+    while True:
+        LOGGER.info("Looking at next task directive without queues...")
+
+        # find
+        try:
+            task_directive = await task_directives_client.find_one(
+                dict(queues=[]),  # has no queues
+                sort=[
+                    ("priority", DESCENDING),  # highest first
+                    ("timestamp", ASCENDING),  # oldest first
+                ],
+            )
+        except db.client.DocumentNotFoundException:
+            LOGGER.info("NOTHING FOR TASK_MQ_ASSEMBLY TO START UP")
+            await asyncio.sleep(ENV.TASK_MQ_ASSEMBLY_SHORT_DELAY)
+            continue
+
+        # request to mqs
+        LOGGER.info(
+            f"REQUESTING {task_directive['n_queues']} queues (task_id={task_directive['task_id']})..."
+        )
+        resp = await mqs_rc.request(
+            "POST",
+            "/mq-group",
+            dict(
+                criteria=dict(
+                    priority=task_directive["priority"],
+                    n_queues=task_directive["n_queues"],
+                )
+            ),
+        )
+        # TODO - Add error handling ^^^
+
+        # update task directive -- insert queues
+        await task_directives_client.find_one_and_update(
+            dict(task_id=task_directive["task_id"]),
+            dict(queues=[p["mqid"] for p in resp["mqprofiles"]]),
```

Comment:
Can you explain how the queues here get mapped to tasks for the TMS to run?  Which ones are input and output?

Reply:
That logic will be in the TMS. I'm planning to inject the queue ids into env vars (`EWMS_PILOT_QUEUE_1`, etc.). Then the pilot uses those accordingly. The most important part is to keep the ordering straight.

I've gone back and forth on whether to have user pass in labels for each queue. There's a `nickname` field for https://github.com/Observation-Management-Service/ewms-message-queue-service/blob/main/Docs/Models/MQProfileObject.md

Reply:
I'm guessing we'll revisit this at some point, when we start doing resource-based queues and there are potentially multiple queues per task.  But maybe not something to worry about now.